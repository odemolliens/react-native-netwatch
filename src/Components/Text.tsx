import * as React from 'react';
import { useContext } from 'react';
import { StyleProp, StyleSheet } from 'react-native';
import { Text as RNPText } from 'react-native-paper';
import { ThemeContext } from '../Theme';

export interface IProps {
  children: string;
  style?: StyleProp<any>;
  numberOfLines?: number;
}

export const Text = (props: IProps) => {
  const theme = useContext(ThemeContext);

  return (
    <RNPText style={[styles.text, { color: theme.textColorOne }, props.style]} {...props}>
      {props.children}
    </RNPText>
  );
};

export const TextSecondaryColor = (props: IProps) => {
  const theme = useContext(ThemeContext);

  return (
    <RNPText style={[styles.text, { color: theme.textColorTwo }, props.style]} {...props}>
      {props.children}
    </RNPText>
  );
};

export const Title = (props: IProps) => {
  const theme = useContext(ThemeContext);

  return (
    <RNPText style={[styles.text, styles.text, { color: theme.textColorOne }, props.style]} {...props}>
      {props.children}
    </RNPText>
  );
};

const styles = StyleSheet.create({
  text: {
    fontSize: 12,
    lineHeight: 20,
  },
  textBold: {
    fontWeight: 'bold',
  },
  subtitle: {
    fontSize: 14,
    lineHeight: 22,
  },
  title: {
    fontSize: 16,
    lineHeight: 24,
  },
});
