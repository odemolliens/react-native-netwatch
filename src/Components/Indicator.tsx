import React, { useState, useRef, useEffect, useContext } from 'react';
import { Animated, TouchableOpacity, StyleSheet } from 'react-native';
import { ThemeContext } from '../Theme';
import { Title } from './Text';

export interface IProps {
  success: number;
  warning: number;
  failed: number;
}

// This component is a bar divided in three parts
// A part to represent success request, a part for warning and a part
// for requests failed.
// It is possible to press on the indicator to expand it and displayed
// a value in percent for each part.
export const Indicator: React.FC<IProps> = (props: IProps) => {
  const theme = useContext(ThemeContext);
  const height = useRef(new Animated.Value(6)).current;
  const [isOpened, setIsOpened] = useState<boolean>(false);

  useEffect(() => {
    Animated.timing(height, {
      toValue: isOpened ? 24 : 6,
      duration: 500,
      useNativeDriver: false,
    }).start();
  }, [isOpened]);

  return (
    <TouchableOpacity testID="openStatsButton" style={{ flexDirection: 'row' }} onPress={() => setIsOpened(!isOpened)}>
      <Animated.View
        key="success"
        style={{
          height,
          width: `${props.success.toString()}%`,
          backgroundColor: theme.successColor,
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        {isOpened && props.success > 10 && (
          <Title style={[styles.text, { color: theme.textColorOne }]}>{`${Math.round(props.success)}%`}</Title>
        )}
      </Animated.View>
      <Animated.View
        key="warning"
        style={{
          height,
          width: `${props.warning.toString()}%`,
          backgroundColor: theme.warningColor,
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        {isOpened && props.warning > 10 && (
          <Title style={[styles.text, { color: theme.textColorOne }]}>{`${Math.round(props.warning)}%`}</Title>
        )}
      </Animated.View>
      <Animated.View
        key="failure"
        style={{
          height,
          width: `${props.failed.toString()}%`,
          backgroundColor: theme.failedColor,
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        {isOpened && props.failed > 10 && (
          <Title style={[styles.text, { color: theme.textColorOne }]}>{`${Math.round(props.failed)}%`}</Title>
        )}
      </Animated.View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  text: {
    fontSize: 16,
    fontWeight: 'bold',
  },
});

export default Indicator;
