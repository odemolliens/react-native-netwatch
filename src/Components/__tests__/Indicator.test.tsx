import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { Indicator, IProps } from '../Indicator';
import { Title } from '../Text';

describe('Indicator test suite', () => {
  let component: ShallowWrapper;
  let props: IProps;
  let useEffect;
  let mockUseEffect;
  const setIsOpened = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    useEffect = jest.spyOn(React, 'useEffect');
    mockUseEffect = () => {
      useEffect.mockImplementationOnce(f => f());
    };

    jest.mock('react', () => ({
      ...jest.requireActual('react'),
      useState: jest.fn(),
    }));
  });

  describe('Tests source', () => {
    it('should render properly', () => {
      givenProps();
      givenComponent();
      expect(component).toMatchSnapshot();
    });

    it('should called setIsOpened', () => {
      const useStateMock: any = isOpened => [isOpened, setIsOpened];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      givenProps();
      mockUseEffect();
      givenComponent();
      whenPressingButton('openStatsButton');
      expect(component).toMatchSnapshot();
      expect(setIsOpened).toHaveBeenCalledTimes(1);
    });

    it('should show full indicator with many failed', () => {
      const useStateMock: any = isOpened => [true, setIsOpened];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      givenProps(5, 5, 90);
      givenComponent();
      whenPressingButton('openStatsButton');
      mockUseEffect();
      expect(component).toMatchSnapshot();
      expect(setIsOpened).toHaveBeenCalledTimes(1);
      expect(component.find(Title)).toHaveLength(1);
      expect(component.find(Title).prop('children')).toEqual('90%');
    });

    it('should show full indicator with many success', () => {
      const useStateMock: any = isOpened => [true, setIsOpened];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      givenProps(75, 20, 5);
      givenComponent();
      whenPressingButton('openStatsButton');
      mockUseEffect();
      expect(component).toMatchSnapshot();
      expect(setIsOpened).toHaveBeenCalledTimes(1);
      expect(component.find(Title)).toHaveLength(2);
      expect(component.find(Title).at(0).prop('children')).toEqual('75%');
      expect(component.find(Title).at(1).prop('children')).toEqual('20%');
      expect(component.find(Title).at(2)).toEqual({});
    });
  });

  // UTILITIES
  function whenPressingButton(testId: string) {
    component.find(`[testID="${testId}"]`).simulate('press');
  }

  // GIVEN
  function givenComponent() {
    component = shallow(<Indicator {...props} />);
  }

  function givenProps(success: number = 34, warning: number = 33, failed: number = 33) {
    props = {
      success,
      warning,
      failed,
    };
  }
});
