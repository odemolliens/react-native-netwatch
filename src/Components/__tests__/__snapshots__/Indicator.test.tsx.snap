// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indicator test suite Tests source should called setIsOpened 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Object {
      "flexDirection": "row",
    }
  }
  testID="openStatsButton"
>
  <ForwardRef(AnimatedComponentWrapper)
    key="success"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#4D7C0F",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "34%",
      }
    }
  />
  <ForwardRef(AnimatedComponentWrapper)
    key="warning"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#D97706",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "33%",
      }
    }
  />
  <ForwardRef(AnimatedComponentWrapper)
    key="failure"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#B91C1C",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "33%",
      }
    }
  />
</ForwardRef>
`;

exports[`Indicator test suite Tests source should render properly 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Object {
      "flexDirection": "row",
    }
  }
  testID="openStatsButton"
>
  <ForwardRef(AnimatedComponentWrapper)
    key="success"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#4D7C0F",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "34%",
      }
    }
  />
  <ForwardRef(AnimatedComponentWrapper)
    key="warning"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#D97706",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "33%",
      }
    }
  />
  <ForwardRef(AnimatedComponentWrapper)
    key="failure"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#B91C1C",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "33%",
      }
    }
  />
</ForwardRef>
`;

exports[`Indicator test suite Tests source should show full indicator with many failed 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Object {
      "flexDirection": "row",
    }
  }
  testID="openStatsButton"
>
  <ForwardRef(AnimatedComponentWrapper)
    key="success"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#4D7C0F",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "5%",
      }
    }
  />
  <ForwardRef(AnimatedComponentWrapper)
    key="warning"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#D97706",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "5%",
      }
    }
  />
  <ForwardRef(AnimatedComponentWrapper)
    key="failure"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#B91C1C",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "90%",
      }
    }
  >
    <Title
      style={
        Array [
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
          },
          Object {
            "color": "#F9FAFB",
          },
        ]
      }
    >
      90%
    </Title>
  </ForwardRef(AnimatedComponentWrapper)>
</ForwardRef>
`;

exports[`Indicator test suite Tests source should show full indicator with many success 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Object {
      "flexDirection": "row",
    }
  }
  testID="openStatsButton"
>
  <ForwardRef(AnimatedComponentWrapper)
    key="success"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#4D7C0F",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "75%",
      }
    }
  >
    <Title
      style={
        Array [
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
          },
          Object {
            "color": "#F9FAFB",
          },
        ]
      }
    >
      75%
    </Title>
  </ForwardRef(AnimatedComponentWrapper)>
  <ForwardRef(AnimatedComponentWrapper)
    key="warning"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#D97706",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "20%",
      }
    }
  >
    <Title
      style={
        Array [
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
          },
          Object {
            "color": "#F9FAFB",
          },
        ]
      }
    >
      20%
    </Title>
  </ForwardRef(AnimatedComponentWrapper)>
  <ForwardRef(AnimatedComponentWrapper)
    key="failure"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#B91C1C",
        "height": Object {
          "interpolate": [Function],
          "setValue": [Function],
        },
        "justifyContent": "center",
        "width": "5%",
      }
    }
  />
</ForwardRef>
`;
